### PR TITLE
Navbar: Main Navigation

### PR DESCRIPTION
**Pull request by balakrishnan-s5**
### **Navbar Component** 
Changes 
1. Added Sub menu to Nav bar items mobile and Desktop
2. Added search component for mobile and Desktop
3. Secondary/Agent logo added to Header and mobile Hamburger Menu

- [**keep**] branch

**Changelog entry :** Nill

- [WARNING ] Description of change in `**Navbar**`

**Additional information**

    You may refer to [this](https://www.figma.com/file/us90jOpWBahsqK2VAluPvD/BookingSG-Design-System?node-id=3033%3A18683&t=kuvaNlzQYxIWZDcn-0)  
    
Any other information you would like to include

[BOOKINGSG-3759](https://jira.ship.gov.sg/browse/BOOKINGSG-3759)

### CODE REVIEW

File: src/navbar/navbar.tsx [hunk 1]
```diff
@@ -179,52 +210,107 @@ const Component = <T,>(
...
+                                    {!toggleheaderClick && (
+                                        <NavbarItems
+                                            items={items.desktop}
+                                            onItemClick={handleNavItemClick}
+                                            selectedId={selectedId}
+                                        />
+                                    )}
+
+                                    <NavElementsRightContainer>
+                                        {children}
```
Comment (qroll) [inline src/navbar/navbar.tsx:247]: what is this `children` intended for? I would prefer a named prop because `children` can be ambiguous

File: src/navbar/navbar.tsx [hunk 2]
```diff
@@ -179,52 +210,107 @@ const Component = <T,>(
...
+                                        compress={compress}
+                                        onClick={handleSecondaryBrandClick}
+                                        data-testid="main__brand"
+                                    />
+                                </NavLogoContainer>
+                            )}
+                        </NavBrandContainer>
+
+                        {!hideNavElements && (
+                            <NavElementsLeftContainer>
```
Comment (qroll) [inline src/navbar/navbar.tsx:236]: seems odd to nest a right container in a left container

File: src/navbar/navbar.tsx [hunk 3]
```diff
@@ -179,52 +210,107 @@ const Component = <T,>(
     const renderNavbar = () => {
```
Comment (qroll) [inline src/navbar/navbar.tsx:210]: this function is getting too long. could it be broken up into smaller render functions with `renderBrand()`, `renderDesktopNavItems()`, `renderDesktopNavActions()` etc.?

File: src/navbar/menu.styles.tsx [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+export const Link = styled(Text.Hyperlink.Small)<StyleProps>`
+    display: flex;
+    position: relative;
+    align-items: left;
+    text-align: left;
+    color: ${Color.Neutral[1]};
+    font-style: normal;
+    font-weight: 400;
+    font-size: 14px !important;
+    line-height: 20px;
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:69]: use `${TextStyleHelper.getTextStyle("H6", "regular")}` instead to override the font styling. avoid using `!important`

**Review by qroll — CHANGES_REQUESTED** [review]
have some initial comments
- amend css to follow DS conventions
- refactor navbar rendering
- search input may need to be reworked slightly (will have more clarity after discussion with UX)

File: src/navbar/drawer.styles.tsx [hunk 5]
```diff
@@ -102,3 +102,48 @@ export const CloseButton = styled(IconButton)`
...
+    height: 100%;
+    flex-direction: row;
+    justify-content: space-between;
+    align-items: center;
+    position: relative;
+`;
+export const NavSeprator = styled.div`
+    display: flex;
+    background-color: ${Color.Neutral[5]};
+    height: 2.8125rem;
```
Comment (keithtxw) [inline src/navbar/drawer.styles.tsx:125]: This is an odd value. 2.8125 translates to 45px. Most of the measurements runs in multiples of 4. Could you clarify the correct value instead?

File: src/navbar/drawer.styles.tsx [hunk 6]
```diff
@@ -102,3 +102,48 @@ export const CloseButton = styled(IconButton)`
...
+
+export const NavLogoContainer = styled.div`
+    display: flex;
+    height: 100%;
+    flex-direction: row;
+    justify-content: space-between;
+    align-items: center;
+    position: relative;
+`;
+export const NavSeprator = styled.div`
```
Comment (keithtxw) [inline src/navbar/drawer.styles.tsx:122]: Typo naming

File: src/navbar/types.ts [hunk 7]
```diff
@@ -60,10 +68,25 @@ export type DrawerDismissalMethod =
 export interface NavbarDrawerProps extends NavbarSharedProps {
     show: boolean;
     children: JSX.Element | JSX.Element[];
-    onClose?: (() => void) | undefined;
+    onClose?: () => void | undefined;
+    resourcesSecondaryBrand?: NavbarResourcesProps | undefined;
+    handleSecondaryBrandClick?: (
```
Comment (keithtxw) [inline src/navbar/types.ts:73]: onSecondaryBrandClick

File: src/navbar/types.ts [hunk 8]
```diff
@@ -60,10 +68,25 @@ export type DrawerDismissalMethod =
 export interface NavbarDrawerProps extends NavbarSharedProps {
     show: boolean;
     children: JSX.Element | JSX.Element[];
-    onClose?: (() => void) | undefined;
+    onClose?: () => void | undefined;
+    resourcesSecondaryBrand?: NavbarResourcesProps | undefined;
```
Comment (keithtxw) [inline src/navbar/types.ts:72]: i think `resources` should be refactored to handle the secondary stuff also. I would imagine inside it will have `primary` and `secondary`

File: src/navbar/types.ts (hunk 7 above)
Comment (keithtxw) [reply src/navbar/types.ts:73]: on side note, should merge into `onBrandClick` but that returns a param to tell whether is it the pri or secondary brand that was clicked

File: stories/navbar/props-table.tsx [hunk 9]
```diff
@@ -17,6 +17,12 @@ const DATA: ApiTableSectionProps[] = [
                     "Navbar resources such as the brand name and logo source",
```
Comment (keithtxw) [inline stories/navbar/props-table.tsx:17]: I think we've mentioned previously that the first PR should cover the component and a basic story first. Stuff like props table should be a separate PR after we're happy with the props and component

File: src/navbar/types.ts [hunk 10]
```diff
@@ -60,10 +68,25 @@ export type DrawerDismissalMethod =
 export interface NavbarDrawerProps extends NavbarSharedProps {
     show: boolean;
     children: JSX.Element | JSX.Element[];
-    onClose?: (() => void) | undefined;
+    onClose?: () => void | undefined;
```
Comment (keithtxw) [inline src/navbar/types.ts:71]: We need the parenthesis else you are saying the function returns a void or undefined

File: src/navbar/types.ts [hunk 11]
```diff
@@ -47,9 +55,9 @@ export interface NavbarSharedProps {
     resources?: NavbarResourcesProps | undefined;
     actionButtons?: NavbarActionButtonsProps | undefined;
     /** Triggered when the brand icon is being clicked */
-    onBrandClick?:
-        | ((event: React.MouseEvent<HTMLAnchorElement>) => void)
-        | undefined;
+    onBrandClick?: (
```
Comment (keithtxw) [inline src/navbar/types.ts:58]: We need the extra parenthesis. Revert change

File: src/navbar/types.ts [hunk 12]
```diff
@@ -8,12 +8,20 @@ export interface NavItemProps<T> extends TextLinkProps {
     id: string;
     "data-testid"?: string | undefined;
     options?: T | undefined;
+    subMenu?: NavSubItemProps<T>[] | undefined;
+}
+
+export interface NavSubItemProps<T> extends TextLinkProps {
```
Comment (keithtxw) [inline src/navbar/types.ts:14]: This should be `NavItemCommonProps` and `NavItemProps` extends this instead

**Review by keithtxw — CHANGES_REQUESTED** [review]

File: src/navbar/search-input.tsx [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+    // =============================================================================
+    const ref = useRef(null);
+
+    // =============================================================================
+    // EFFECTS
+    // =============================================================================
+    useEffect(() => {
+        setShowDropdown(0);
+        const handleClickOutside = (event) => {
+            if (ref.current && !ref.current.contains(event.target)) {
```
[the remark this answers is not in the document]
Comment (qroll) [reply src/navbar/search-input.tsx:73]: correcting myself, got confused by the behaviour as ref was assigned in multiple places

File: src/navbar/brand.styles.tsx [hunk 14]
```diff
@@ -27,4 +27,20 @@ export const Clickable = styled.a<StyleProps>`
...
+        margin-right: 1rem;
+    }
+
+    ${MediaQuery.MaxWidth.mobileM} {
+        margin-right: 1rem;
+    }
+
+    ${MediaQuery.MaxWidth.mobileS} {
+        margin-right: 1rem;
+    }
```
[the remark this answers is not in the document]
Comment (balakrishnan-s5) [reply src/navbar/brand.styles.tsx]: duplicate media queries Removed

File: src/navbar/drawer.styles.tsx [hunk 15]
```diff
@@ -102,3 +102,48 @@ export const CloseButton = styled(IconButton)`
...
+    display: flex;
+    background-color: ${Color.Neutral[5]};
+    height: 2.8125rem;
+    width: 0.05rem;
```
[the remark this answers is not in the document]
Comment (balakrishnan-s5) [reply src/navbar/drawer.styles.tsx:126]: rem value changed to PX

File: src/navbar/drawer.styles.tsx [hunk 16]
```diff
@@ -102,3 +102,48 @@ export const CloseButton = styled(IconButton)`
...
+    left: 13.435625rem;
```
[the remark this answers is not in the document]
Comment (balakrishnan-s5) [reply src/navbar/drawer.styles.tsx]: Changed the size

File: src/navbar/drawer.styles.tsx [hunk 17]
```diff
@@ -102,3 +102,48 @@ export const CloseButton = styled(IconButton)`
...
+    margin-right: 1.504375rem;
```
[the remark this answers is not in the document]
Comment (balakrishnan-s5) [reply src/navbar/drawer.styles.tsx]: Changed the size

File: src/navbar/search-input.tsx [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+}: Props<T>): JSX.Element => {
+    // =============================================================================
+    // STATE DECLARATIONS
+    // =============================================================================
+    const [toggleInput, setToggleInput] = useState<boolean>(false);
+    const [inputValue, setInputValue] = useState<string>("");
+    const [itemsLocal, setItemsLocalValue] = useState<NavSubItemProps<T>[]>([]);
+    const [toggleDropdown, setToggleDropdown] = useState<boolean>(false);
+    const [toggleInputFocus, setToggleInputFocus] = useState<boolean>(false);
+    const [showDropdown, setShowDropdown] = useState<number>(0);
```
[the remark this answers is not in the document]
Comment (balakrishnan-s5) [reply src/navbar/search-input.tsx]: Changed

File: src/navbar/search-input.tsx [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+    selectedId?: string | undefined;
+    /** toggle for mobile or desktop view */
+    mobile?: boolean | undefined;
+    toggleheaderClick?: boolean | undefined;
+    onItemClick: (
+        event: React.MouseEvent<HTMLAnchorElement>,
+        item: NavSubItemProps<T>
+    ) => void;
+    onToggleHeaderClick: () => void;
+}
```
[the remark this answers is not in the document]
Comment (balakrishnan-s5) [reply src/navbar/search-input.tsx:37]: Modified accordingly all 3

File: src/navbar/drawer.tsx [hunk 20]
```diff
@@ -51,11 +62,25 @@ const Component = (
...
+                    data-testid="drawer__brand"
+                />
+                {resourcesSecondaryBrand && (
+                    <NavLogoContainer>
+                        <NavSeprator />
+                        <Brand
+                            resources={resourcesSecondaryBrand}
+                            compress={true}
+                            onClick={handleSecondaryBrandClick}
+                            data-testid="main__brand"
```
[the remark this answers is not in the document]
Comment (balakrishnan-s5) [reply src/navbar/drawer.tsx]: Changed

File: src/navbar/search-input.tsx [hunk 21]
```diff
@@ -0,0 +1,311 @@
...
+    // =============================================================================
+    // STATE DECLARATIONS
+    // =============================================================================
+    const [toggleInput, setToggleInput] = useState<boolean>(false);
+    const [inputValue, setInputValue] = useState<string>("");
+    const [itemsLocal, setItemsLocalValue] = useState<NavItemCommonProps<T>[]>(
+        []
+    );
+    const [toggleDropdown, setToggleDropdown] = useState<boolean>(false);
+    const [showDropdown, setShowDropdown] = useState<string>("on-blur");
```
Comment (qroll) [inline src/navbar/search-input.tsx:58]: as discussed
- can simplify this to a `focused` (meaning either blue outline or default grey) and `expanded` state
- can try to reuse `InputSelectWrapper`. for the case where input is greyed out when there are no search results, you can override the default `:focus-within` style

**Review by qroll — CHANGES_REQUESTED** [review]
as discussed,
- to update px to rem
- remove redundant media queries
- refactor select-input

File: src/navbar/navbar-items.tsx [hunk 22]
```diff
@@ -52,30 +121,72 @@ export const NavbarItems = <T,>({
...
+                        {renderItems(mobile)}
+                        {items &&
+                            selectedIndex >= 0 &&
+                            items[selectedIndex] &&
+                            toggleDropdown && (
+                                <Menu
+                                    items={items[selectedIndex].subMenu}
+                                    selectedId={selectedId}
+                                    mobile={false}
+                                    onItemClick={handleSubLinkClick(
```
Comment (qroll) [inline src/navbar/navbar-items.tsx:182]: actually since menu also has an `onItemClick` that passes the event and selected item, could `handleSubLinkClick` simply take in those arguments instead?

File: src/navbar/menu.styles.tsx [hunk 23]
```diff
@@ -0,0 +1,207 @@
...
+    ${MediaQuery.MaxWidth.tablet} {
+        width: 100%;
+
+        :not(:last-of-type) {
+            margin-right: 0;
+        }
+    }
+`;
+
+export const Link = styled(Text.Hyperlink.Small)<StyleProps>`
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:60]: there are a lot of redundant css styles here. for example line-clamps are the same on desktop and tablet, so there's no need to repeat them again

similarly, `TextStyleHelper` already covers font-size, font-weight, line-height and letter-spacing properties, so there's no need to specify them again

File: src/navbar/menu.styles.tsx [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+    left: 0;
+    top: 102%;
+    min-width: 250px;
+    /* N8 #FFFFFF */
+    position: absolute;
+    overflow-x: hidden;
+
+    /* N8 #FFFFFF */
+    background: #ffffff;
+    box-shadow: 0px 2px 8px rgba(75, 83, 159, 0.24);
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:147]: use built-in theme colors instead of hardcoding the hexes or rgba values

usually the Figma comments will tell you which color to use. for example N8 means `Color.Neutral[8]`

File: src/navbar/navbar-items.tsx [hunk 25]
```diff
@@ -26,22 +32,85 @@ export const NavbarItems = <T,>({
...
+            setShowDrawer(index);
+            onItemClick(event, item[index]);
+            setToggleDropdown(false);
+        };
+    };
+
     // =============================================================================
     // RENDER FUNCTIONS
     // =============================================================================
     const renderItems = (isMobile = false) => {
```
Comment (qroll) [inline src/navbar/navbar-items.tsx:110]: it might be worth splitting up the rendering of mobile and desktop components

File: src/navbar/menu.styles.tsx [hunk 26]
```diff
@@ -0,0 +1,188 @@
...
+    margin-bottom: 0.5rem;
+
+    :first-child {
+        margin-top: 1rem;
+    }
+
+    :last-child {
+        margin-bottom: 1rem;
+    }
+    :not(:firt-of-type) {
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:154]:
```suggestion
    :not(:first-of-type) {
```

File: src/navbar/menu.styles.tsx [hunk 27]
```diff
@@ -0,0 +1,193 @@
...
+        width: 0.25rem;
+        background-color: ${Color.Accent.Light[1]};
+    }
+`;
+
+export const MenuBar = styled.ul`
+    width: 100%;
+    overflow: auto;
+    display: grid;
+    margin-top: 0rem;
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:123]: can omit the unit and leave as `0`. trouble you to update all other instances

File: src/navbar/menu.styles.tsx [hunk 28]
```diff
@@ -0,0 +1,193 @@
...
+    width: 100%;
+
+    ${MediaQuery.MaxWidth.tablet} {
+        display: none;
+    }
+`;
+export const MenuTopSpace = styled.div`
+    width: 100%;
+    height: 0.5rem;
+`;
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:193]: seem to be unused. please look through and remove unused code

File: src/navbar/navbar-items.styles.tsx [hunk 29]
```diff
@@ -39,9 +46,8 @@ export const MobileWrapper = styled.ul`
 // =============================================================================
 export const LinkItem = styled.li`
     position: relative;
-    display: flex;
+    display: contents;
```
Comment (qroll) [inline src/navbar/navbar-items.styles.tsx:49]: can this be reverted to `flex`? would prefer to avoid `contents` as it makes it harder to reason about the css + has poor accessibility support in older browsers. can try moving LinkIndicator as a sibling of MenuItemContainer instead to get the absolute positioning

File: src/navbar/menu.styles.tsx [hunk 30]
```diff
@@ -0,0 +1,162 @@
...
+    margin-right: 0;
+
+    :first-child {
+        padding-top: 1rem;
+        padding-bottom: 0.5rem;
+    }
+
+    :not(:first-of-type) {
+        padding-top: 0.5rem;
+        padding-bottom: 0.5rem;
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:136]: the pseudo-class is not needed here, can set this as a default padding

can also simplify the expression to `padding: 0.5rem 0`

File: src/navbar/menu.styles.tsx [hunk 31]
```diff
@@ -0,0 +1,162 @@
...
+    width: 100%;
+    position: relative;
+    display: flex;
+    align-items: left;
+    margin-left: 0;
+    margin-right: 0;
+
+    :first-child {
+        padding-top: 1rem;
+        padding-bottom: 0.5rem;
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:131]: following the previous change, this padding bottom is not needed

File: src/navbar/menu.styles.tsx [hunk 32]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    :not(:first-of-type) {
+        padding-top: 0.5rem;
+        padding-bottom: 0.5rem;
+    }
+
+    :last-child {
+        padding-top: 0.5rem;
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:140]: following the previous change, this padding top is not needed

File: src/navbar/menu.styles.tsx [hunk 33]
```diff
@@ -0,0 +1,162 @@
...
+// =============================================================================
+// LINK ITEMS
+// =============================================================================
+export const MenuItem = styled.li`
+    width: 100%;
+    position: relative;
+    display: flex;
+    align-items: left;
+    margin-left: 0;
+    margin-right: 0;
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:127]: doesn't seem to be needed

File: src/navbar/menu.tsx [hunk 34]
```diff
@@ -0,0 +1,84 @@
...
+                    <Link
+                        data-testid={testId}
+                        weight={textWeight}
+                        $selected={selected} /* for mobile */
+                        {...otherItemAttrs}
+                        onClick={handleLinkClick(item)}
+                        {...options}
+                    >
+                        {children}
+                        {selected && <LinkIndicator />}
```
Comment (qroll) [inline src/navbar/menu.tsx:67]: is there a scenario where submenu item has selected styling?

File: src/navbar/menu.styles.tsx [hunk 35]
```diff
@@ -0,0 +1,162 @@
...
+    display: flex;
+    position: relative;
+    align-items: left;
+    text-align: left;
+    color: ${Color.Neutral[1]};
+
+    padding-left: 1rem;
+    padding-right: 1rem;
+    padding-top: 1px;
+    padding-bottom: 1px;
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:56]: can simplify to a single `padding` property. also is the 1px needed or can it just be 0?

File: src/navbar/menu.styles.tsx [hunk 36]
```diff
@@ -0,0 +1,162 @@
...
+
+    overflow: hidden;
+    display: -webkit-box;
+    -webkit-box-orient: vertical;
+    -webkit-line-clamp: 2; /* start showing ellipsis when 2rd line is reached */
+    white-space: pre-wrap; /* let the text wrap preserving spaces */
+
+    /* identical to box height, or 186% */
+    letter-spacing: 0.0075rem;
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:65]: to remove

File: src/navbar/menu.styles.tsx [hunk 37]
```diff
@@ -0,0 +1,162 @@
...
+        padding-bottom: 1rem;
+    }
+`;
+
+// =============================================================================
+// LINK ITEMS
+// =============================================================================
+
+export const Link = styled(Text.Hyperlink.Small)<StyleProps>`
+    ${TextStyleHelper.getTextStyle("H6", 400)};
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:45]:
```suggestion
    ${TextStyleHelper.getTextStyle("H6", "regular")};
```

File: src/navbar/menu.styles.tsx [hunk 38]
```diff
@@ -0,0 +1,162 @@
...
+    width: 100%;
+    display: flex;
+    position: relative;
+    align-items: left;
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:49]: can be removed. anyway there's no `left` value for `align-items`, did you mean flex-start?

File: src/navbar/menu.styles.tsx [hunk 39]
```diff
@@ -0,0 +1,162 @@
...
+    -webkit-box-orient: vertical;
+    -webkit-line-clamp: 2; /* start showing ellipsis when 2rd line is reached */
+    white-space: pre-wrap; /* let the text wrap preserving spaces */
+
+    /* identical to box height, or 186% */
+    letter-spacing: 0.0075rem;
+
+    :active,
+    :focus {
+        color: ${Color.Primary} !important;
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:69]: avoid using `!important`

File: src/navbar/menu.styles.tsx [hunk 40]
```diff
@@ -0,0 +1,162 @@
+import styled from "styled-components";
+import { Color } from "../color";
+import { MediaQuery } from "../media";
+import { Text, TextStyleHelper } from "../text";
+import { DesignToken } from "src/design-token";
```
Comment (qroll) [inline src/navbar/menu.styles.tsx:5]: can trouble you to change to a relative import instead

File: src/navbar/menu.tsx [hunk 41]
```diff
@@ -64,7 +64,7 @@ export const Menu = <T,>({
                         {...options}
                     >
                         {children}
-                        {selected && <LinkIndicator />}
+                        {isMobile && selected && <LinkIndicator />}
```
Comment (qroll) [inline src/navbar/menu.tsx:67]: actually I still don't see a case where this `LinkIndicator` is required. the dark blue border is already set on the `MobileWrapper`

File: src/navbar/navbar-items.styles.tsx [hunk 42]
```diff
@@ -110,19 +106,17 @@ export const LinkIconContainer = styled.div`
...
+    margin: auto 0.25rem auto 0;
 `;
 
-export const UpIcon = styled(ChevronUpIcon)`
-    height: 1.25rem;
-    width: 1.25rem;
-
+export const ChevronIcon = styled(ChevronUpIcon)`
+    height: 1rem;
+    width: 1rem;
```
Comment (qroll) [inline src/navbar/navbar-items.styles.tsx:118]: this should be 1.25rem based on figma

the padding to the right edge should also be 1.375rem

**Review by qroll — COMMENTED** [review]
something else I missed out on previously: the close button position in drawer needs adjustment to be 20px from the right edge

**Review by qroll — APPROVED** [review]
